core/src/main/java/org/apache/calcite/plan/AbstractRelOptPlanner.java

```diff
@@ -197,6 +198,10 @@ public boolean isRuleExcluded(RelOptRule rule) {
     return this;
   }
 
+  @Override public void addExtraMaterializationRules(List<UnifyRule> rules) {
```

Comment:
The materialized recognition rules currently implemented are quite general (based on pure relational algebra). I think adding some additional rules for the materialized recognition of specific scenes will help us to solve the materialized recognition of specific scenes, rather than adding them all together. Even if no additional rules are added, the current physical and chemical recognition rules, such as `CalcToCalcUnifyRule` and `JoinOnCalcsToJoinUnifyRule`, are still the most basic materialized recognition rules, and we should keep them unchanged.I think it's better to retain the most basic materialized recognition rules and add custom materialized recognition rules at the same time.

Reply:
Yes, I understand that they are quite general. My point was that since we are already changing the interface, we could use the same method to add any materialization rules, allowing full control over which ones are enabled throughout the planning. However, I am fine with this approach if others think it is fine.

Reply:
I also share @jcamachor view point. If we decide to allow customization of this part its better to allow full-control. 

Furthermore, to be consistent with the existing APIs of the planner, I would opt for a per rule addition:
`addMaterializationRule(UnifyRule rule)`

Going one step further, is it really necessary to introduce a new API for this? The point is that we want to add new rules to the planner so can't we use the existing `addRule(RelOptRule)` interface? I guess we want to distinguish those rules that are specific to materialization but this could be done in the implementation of `addRule`.

What do you think?

Reply:
@jcamachor @zabetak
Thank you very much. I have update the code. The rule of materialized view is `AbstractUnifyRule`, not `RelOptRule`. I suggest registering materialized view recognition algorithm through `addMaterializationRules(List<UnifyRule> rules)`. This interface can also be compatible with `UnifyRule`. If a single 'unifyrule' is registered, it is ok. In the project, we call ` org.apache.calcite . plan.RelOptMaterializations#useMaterializedViews () ' interface for view recognition. Hope to get feedback, thank you again.

Reply:
@jcamachor @zabetak
Similar to this PR, register custom materialized view normalization algorithm code[1]. Please help review when you have time. Thanks.
[1] [Add an interface in RelOptMaterializations to allow registering normalization rules](https://github.com/apache/calcite/pull/2262)

Reply:
@jcamachor @zabetak 
@xy2953396112 has updated the code according to the comments.
Any more comments?

Reply:
@hsyuan I added a few comments in the JIRA but don't have much time to follow up on this. I would be happy if someone else can take over.

Reply:
@zabetak 
Thank you for spending a long time on this pr.

Reply:
@hsyuan 
I have updated the code.Avoid changing the `RelOptPlanner` APIs,and It can be compatible with the original implementation.Please help review.